lib/ts/ArgParser.cc

```diff
@@ -0,0 +1,506 @@
+    std::cout << "\nOptions: " << std::endl;
+    for (auto it : command._option_list) {
+      std::string msg = "  " + it.first;
+      if (!it.second._opt_key.empty()) {
+        msg = msg + ", " + it.second._opt_key;
+      }
+      // create the indent for output
+      std::string indent;
+      for (unsigned long i = 0; i < 30 - msg.size(); i++) {
+        indent += " ";
+      }
+      std::cout << msg << ": " << indent << it.second._opt_description << std::endl;
+    }
+  }
+  // 4. output example usage
+  if (!command._example_usage.empty()) {
+    std::cout << "\nExample Usage: " << command._example_usage << std::endl;
+  }
+  exit(0);
+}
+
+void
+ArgParser::version_message() const
+{
+  // unified version message of ATS
+  AppVersionInfo appVersionInfo;
+  appVersionInfo.setup(PACKAGE_NAME, _argv[0].c_str(), PACKAGE_VERSION, __DATE__, __TIME__, BUILD_MACHINE, BUILD_PERSON, "");
+  ink_fputln(stdout, appVersionInfo.FullVersionInfoStr);
+  exit(0);
+}
+
+// Top level call of parsing
+ParsedArgs
+ArgParser::parse(const char **argv)
+{
+  // deal with argv first
+  int size = 0;
+  _argv.clear();
+  while (argv[size]) {
+    _argv.push_back(argv[size]);
+    size++;
+  }
+  if (size == 0) {
+    std::cout << "Error: invalid argv provided" << std::endl;
+    exit(1);
+  }
+  // the name of the program only
+  _argv[0]                 = _argv[0].substr(_argv[0].find_last_of('/') + 1);
+  _top_level_command._name = _argv[0];
+
+  ParsedArgs ret; // the parsed arg object to return
+  StringArray args = _argv;
+  // call the recrusive parse method in Command
+  _top_level_command.parse(*this, ret, args);
+  // if there is anything left, then output usage
+  if (!args.empty()) {
+    std::string msg = "Unkown command, option or args:";
+    for (unsigned long i = 0; i < args.size(); i++) {
+      msg = msg + " '" + args[i] + "'";
+    }
+    std::cout << msg << std::endl;
+    help_message();
+  }
+  return ret;
+}
+
+//=========================== Command class ================================
+ArgParser::Command::Command() {}
+
+ArgParser::Command::~Command() {}
+
+ArgParser::Command::Command(std::string const &name, std::string const &description, std::string const &envvar, int arg_num,
+                            Function const &f)
+  : _name(name), _description(description), _arg_num(arg_num), _envvar(envvar), _f(f)
+{
+}
+
+// check if this is a valid option before adding
+void
+ArgParser::Command::check_option(std::string const &name, std::string const &key) const
+{
+  if (name.size() < 3 || name[0] != '-' || name[1] != '-') {
+    // invalid name
+    std::cerr << "Error: invalid long option added: '" + name + "'";
+    exit(1);
+  }
+  if (key.size() > 0 && key[0] != '-') {
+    // invalid key
+    std::cerr << "Error: invalid short option added: '" + key + "'";
+    exit(1);
+  }
+  // find if existing in option list
+  if (_option_list.find(name) != _option_list.end() || _option_map.find(key) != _option_map.end()) {
```

Comment:
Why is there a list _and* a map?

Reply:
`_option_list` is for the name(long option) and `Option` object.
`_option_map`is the map for the key(short option) and name(long option).
It is there for fast searching